src/cls/rgw/cls_rgw_types.h

```diff
@@ -510,6 +520,40 @@ struct rgw_cls_bi_entry {
 };
 WRITE_CLASS_ENCODER(rgw_cls_bi_entry)
 
+// constructed base on index entry read in processing reshard log
+struct rgw_cls_bi_process_log_entry {
+  std::string idx; // whole name of an omap key
+  bool exists;
+  rgw_cls_bi_entry bi_entry;
```

Comment:
rather than add new cls types (`rgw_cls_bi_process_log_entry` and `rgw_reshard_log_entry`) and new functions (`cls_rgw_reshard_log_list` and `cls_rgw_bi_process_log_put`), what do you think about just adding a `uint64_t gen` argument to the existing `cls_rgw_bi_list()` function?

i think that would really help to unify a lot of logic between the two reshard stages. the first stage uses `cls_rgw_bi_list()` to list entries, and `cls_rgw_bi_put()` to write each to the target shard

i'm imagining a second stage that follows that same pattern, except that `cls_rgw_bi_list()` reads entries from the new `BI_BUCKET_RESHARD_LOG_INDEX` namespace for the given `gen`

most of the special handling could then be hidden by cls_rgw.cc:
* `cls_rgw_bi_list()` can trim the BI_BUCKET_RESHARD_LOG_INDEX prefix from `idx` like `reshard_log_list_entries()` currently does,
* deleted keys an be returned in `rgw_cls_bi_entry` as a new `BIIndexType` like `ReshardDeleted`,
* `rgw_bi_put_op()` would handle the deleted entries with `cls_cxx_map_remove_key()` like `rgw_bi_process_log_put_op()` currently does

Reply:
> Otherwise, I have to add a is_delete variable to `rgw_bucket_dir_entry` and `rgw_bucket_olh_entry`, so we can record the {rgw_obj_key}. Which means, even is deleted, the `rgw_bucket_dir_entry` or `rgw_bucket_olh_entry` still need to be created.

we can preserve the original key for `rgw_cls_bi_entry::get_info()` by encoding it in a new `struct rgw_bucket_deleted_entry` for entries with type `BIIndexType::ReshardDeleted`:
```diff
diff --git a/src/cls/rgw/cls_rgw_types.cc b/src/cls/rgw/cls_rgw_types.cc
index ccf92c3765f..e5604cab514 100644
--- a/src/cls/rgw/cls_rgw_types.cc
+++ b/src/cls/rgw/cls_rgw_types.cc
@@ -393,6 +393,12 @@ bool rgw_cls_bi_entry::get_info(cls_rgw_obj_key *key,
     *key = entry.key;
     return false;
   }
+  if (type == BIIndexType::ReshardDeleted) {
+    rgw_bucket_deleted_entry entry;
+    decode(entry, iter);
+    *key = std::move(entry.key);
+    return false;
+  }
 
   rgw_bucket_dir_entry entry;
   decode(entry, iter);
```
this way we don't add any space overhead to `rgw_bucket_dir_entry` or `rgw_bucket_olh_entry`

for the omap entries we store under the BI_BUCKET_RESHARD_LOG_INDEX namespace, we still need a way to differentiate deleted entries from the normal ones with the same omap key. you're currently encoding those as `rgw_reshard_log_entry`, which carries enough information for `cls_rgw_bi_list()` to reconstruct the `rgw_cls_bi_entry`

but given `BIIndexType::ReshardDeleted` and `rgw_bucket_deleted_entry`, `rgw_reshard_log_entry` could contain just the type and bufferlist

we might simplify that even further by storing `rgw_cls_bi_entry` directly for the reshard logs. the only additional field that has is `std::string idx`, which we could leave empty until `cls_rgw_bi_list()` is ready to return it to the client

Reply:
> * `rgw_bi_put_op()` would handle the deleted entries with `cls_cxx_map_remove_key()` like `rgw_bi_process_log_put_op()` currently does

i think this part is problematic for upgrades too. we have a strategy to check that source shard osds upgraded, but this expects new behavior from _target_ shard osds. it's probably better for radosgw to handle deleted keys with `librados::ObjectWriteOperation::omap_rm_keys()` instead of passing them to `rgw_bi_put_op()`

Reply:
> it's probably better for radosgw to handle deleted keys with `librados::ObjectWriteOperation::omap_rm_keys()` instead of passing them to `rgw_bi_put_op()`

`cls_rgw_bi_get_vals()` and `rgw_bucket_update_stats()` also need osds upgraded. It may is not easy handling `rgw_bucket_update_stats()` in radosgw. If not easy, can we judge the target osds too, and back to original resharding when there are old osds?

Reply:
> can we judge the target osds too

for the target shards, init_target_index() could try a cls_rgw_bucket_init_index2() call